subprojects/core/src/integTest/groovy/org/gradle/api/internal/changedetection/state/TaskTypeUpToDateIntegrationTest.groovy

```diff
@@ -176,6 +180,33 @@ class TaskTypeUpToDateIntegrationTest extends AbstractIntegrationSpec {
         then: skipped(":copy")
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/9723")
+    def "declaring a task action receiving #incrementalChangesType without declaring outputs is deprecated"() {
+        def input = file('input.txt').createFile()
+        buildFile << """
+            class IncrementalTask extends DefaultTask {
+                @InputFile File input
+             
+                @TaskAction execute(${incrementalChangesType} inputChanges) {
+                }
+            }   
+            
+            task noOutput(type: IncrementalTask) {
+                input = file('${input.name}')
+            }
+        """
+
+        when:
+        executer.expectDeprecationWarning()
+        run 'noOutput'
+        then:
+        outputContains("Using the incremental task API without declaring any outputs has been deprecated. This is scheduled to be removed in Gradle 6.0. Please declare output files for your task or use `task.upToDateWhen { true }`.")
```

Comment:
Should we say `upToDateWhen { true }` or `upToDateWhen()`? I think the latter is better.

Reply:
Using `upToDateWhen()` now.